### PR TITLE
Verif / handle concurrent start

### LINKED ISSUE #794: Verification DM /  Handle concurrent .start after .ready
After the `m.key.verification.ready` event is sent, either party can send an
`m.key.verification.start` event to begin the verification.  If both parties
send an `m.key.verification.start` event, and they both specify the same
verification method, then the event sent by the user whose user ID is the
smallest is used, and the other `m.key.verification.start` event is ignored.
In the case of a single user verifying two of their devices, the device ID is
compared instead.  If both parties send an `m.key.verification.start` event,
but they specify different verification methods, the verification should be
cancelled with a `code` of `m.unexpected_message`.

### PR DESCRIPTION
**Pull request by BillCarsonFr**
Fixes #794
(Verification DM / Handle concurrent .start after .ready)

### CODE REVIEW

File: matrix-sdk-android/src/androidTest/java/im/vector/matrix/android/internal/crypto/gossiping/KeyShareTests.kt [hunk 1]
```diff
@@ -90,10 +87,10 @@ class KeyShareTests : InstrumentedTest {
 
         var outGoingRequestId: String? = null
 
-        retryPeriodicallyWithLatch(waitLatch) {
+        mTestHelper.retryPeriodicallyWithLatch(waitLatch) {
             aliceSession2.cryptoService().getOutgoingRoomKeyRequest()
                     .filter { req ->
-                        // filter out request that was known before
+                        // filter out request thwat was known before
```
Comment (onurays) [inline matrix-sdk-android/src/androidTest/java/im/vector/matrix/android/internal/crypto/gossiping/KeyShareTests.kt]: Typo?

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/crypto/verification/DefaultVerificationService.kt [hunk 2]
```diff
@@ -457,7 +457,29 @@ internal class DefaultVerificationService @Inject constructor(
         Timber.d("## SAS onStartRequestReceived ${startReq.transactionId}")
         if (checkKeysAreDownloaded(otherUserId!!, startReq.fromDevice) != null) {
```
Comment (onurays) [inline matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/crypto/verification/DefaultVerificationService.kt]: It seems dangerous to force a nullable field.

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/crypto/verification/DefaultVerificationService.kt [hunk 3]
```diff
@@ -457,7 +457,29 @@ internal class DefaultVerificationService @Inject constructor(
...
+            var existing = getExistingTransaction(otherUserId, tid)
+
+            // After the m.key.verification.ready event is sent, either party can send an
+            // m.key.verification.start event to begin the verification. If both parties
+            // send an m.key.verification.start event, and they both specify the same
+            // verification method, then the event sent by the user whose user ID is the
+            // smallest is used, and the other m.key.verification.start event is ignored.
+            // In the case of a single user verifying two of their devices, the device ID is
+            // compared instead .
+            if (existing != null && !existing.isIncoming) {
```
Comment (onurays) [inline matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/crypto/verification/DefaultVerificationService.kt]: I would prefer to use `existing?.isIncoming == false`

Comment (BillCarsonFr) [reply matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/crypto/verification/DefaultVerificationService.kt]: lint asks then to simplify to !

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/crypto/verification/DefaultVerificationService.kt (hunk 2 above)
Comment (BillCarsonFr) [reply matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/crypto/verification/DefaultVerificationService.kt]: done

File: matrix-sdk-android/src/androidTest/java/im/vector/matrix/android/internal/crypto/gossiping/KeyShareTests.kt (hunk 1 above)
Comment (BillCarsonFr) [reply matrix-sdk-android/src/androidTest/java/im/vector/matrix/android/internal/crypto/gossiping/KeyShareTests.kt]: yep, fixed

**Comment by bmarty** [comment]
Late LGTM :)

**Pull request merged**